rfcs/004 - Better Policy UX RFC.md

```diff
@@ -0,0 +1,156 @@
+                policy_id -- nullable if policy_document is provided
+            ) values (
+                      'some_bucket',
+                      NULL,
+                      iasql_create_policy( -- returns policy ID after creation
+                          'some_policy_name',
+                          '2012-10-17', -- nullable, defaults to 2012-10-17
+                          iasql_create_policy_statement -- returns policy statement ID after creation
+                            '<policy_id>',
+                            'allow',
+                            'aws:arn:123456789012:bucket-resource',
+                            's3:GetObjects', -- comma separated list of actions
+                            '123456789012' -- comma separated list of principals
+                            ), -- Statement 1
+                          iasql_create_policy_statement(...), -- Statement 2
+                          ...
+                          iasql_create_policy_statement(...) -- Statement N
+                    );
+```
+
+### ER diagram for s3 example
+![Proposed model for s3](./assets/004/s3_example.png)
+
+And for Updating, we can equally have something like this:
+
+```sql
+UPDATE bucket set policy_id = iasql_create_policy( -- returns policy ID after creation
+                              'some_policy_name',
+                              '2012-10-17', -- nullable, defaults to 2012-10-17
+                              iasql_create_policy_statement -- returns policy statement ID after creation
+                                '<policy_id>',
+                                'allow',
+                                'aws:arn:123456789012:bucket-resource',
+                                's3:GetObjects', -- comma separated list of actions
+                                '123456789012' -- comma separated list of principals
+                              ), -- Statement 1
+                              iasql_create_policy_statement(...), -- Statement 2
+                              ...
+                              iasql_create_policy_statement(...) -- Statement N
+                        ) WHERE id = '<bucket_id>';
+```
+
+These functions can be used to create policies for any AWS service (SQS, VPC, SNS etc.). We can also add more functions to create policies for
+services in other clouds like GCP, Azure (out of scope).
+
+Now, once the JSON blob is generated from the `Policy` and `Statement` tables, the data in these tables are no longer needed.
+We can choose to keep them but the tables will keep growing and will span billions of records in no time. I recommend we
+have a retention policy for the tables (perhaps we can retain them for 12-24 hours and then delete).
+We can also choose to delete them immediately after the JSON blob is generated, though a periodic bulk-delete is cheaper.
+
+
+### Alternatives Considered
+
+* Creating helper postgres functions to help the user generate policies from scratch (or from existing policies). 
```

Comment:
i believe that this could be a good approach for the user . Using https://awspolicygen.s3.amazonaws.com/policygen.html as a base, have some stored procedure that accepts some parameters, and returns the generated json ? that can be called easily in inserts: INSERT INTO x (policy_value) SELECT generate_policy('type', vals...)

Reply:
Thanks for the suggestion. I considered this one. I will add details as to why I don't think it may be better that the proposed solution.